elyra/pipeline/component_registry.py

```diff
@@ -0,0 +1,214 @@
+                              extension=".r")}
+
+    def __init__(self, component_registry_location: str, parser: ComponentParser):
+        super().__init__()
+        self._component_registry_location = component_registry_location
+        self._parser = parser
+        self.log.info(f'Creating new registry using {self.registry_location}')
+
+    @property
+    def registry_location(self) -> str:
+        return self._component_registry_location
+
+    def get_all_components(self) -> List[Component]:
+        """
+        Retrieve all components from the component_id registry
+        """
+        components: List[Component] = list()
+
+        # Read component_id catalog to get JSON
+        component_entries = self._read_component_registry()
+
+        for component_entry in component_entries:
+            # Parse component_id details and add to list
+            component = self._parser.parse(component_entry)
+            if component:
+                components.extend(component)
+
+        return components
+
+    def get_component(self, component_id) -> Component:
+        """
+        Return the properties JSON for a given component_id.
+        """
+        # Read component_entry catalog to get component_entry with given id
+        adjusted_id = self._parser.get_adjusted_component_id(component_id)
+        component_entry = self._get_component_registry_entry(adjusted_id)
+
+        # Assign adjusted id for the use of parsers if prefixes have been added
+        if adjusted_id != component_id:
+            component_entry.adjusted_id = component_id
+
+        component = self._parser.parse(component_entry)[0]
+        return component
+
+    @staticmethod
+    def get_generic_components() -> List[Component]:
+        return list(ComponentRegistry._generic_components.values())
+
+    @staticmethod
+    def get_generic_component(component_id: str) -> Component:
+        return ComponentRegistry._generic_components.get(component_id)
+
+    @staticmethod
+    def to_canvas_palette(components: List[Component]) -> dict:
+        """
+        Converts registry components into appropriate canvas palette format
+        """
+        # Load jinja2 template
+        loader = PackageLoader('elyra', 'templates/components')
+        template_env = Environment(loader=loader)
+        template = template_env.get_template('canvas_palette_template.jinja2')
+
+        return template.render(components=components)
+
+    @staticmethod
+    def to_canvas_properties(component: Component) -> dict:
+        """
+        Converts registry components into appropriate canvas properties format
+        """
+        loader = PackageLoader('elyra', 'templates/components')
+        template_env = Environment(loader=loader)
+
+        # If component_id is one of the generic set, render with generic template,
+        # else render with the runtime-specific property template
+        if component.id in ('notebooks', 'python-script', 'r-script'):
```

Comment:
Nit: not future proof. Should use key lookup  in `_generic_components` instead of hard-coding them here.

Reply:
we will enhance this in future prs